shark_turbine/kernel/wave/minimize_global_loads.py

```diff
@@ -0,0 +1,237 @@
+    """
+    thread_ids = [THREAD_0, THREAD_1, THREAD_2]
+    new_index = {key: index[key].subs({t: 0 for t in thread_ids}) for key in index}
+    nd_index = delinearize_index(thread_id, shape)
+    for i, key in enumerate(index.keys()):
+        new_index[key].start += nd_index[i]
+        new_index[key].size = load_elems_per_thread
+    return new_index
+
+
+def remove_global_indexing(
+    index: dict[IndexSymbol, IndexSequence], tilingConstraints: list[TilingConstraint]
+) -> dict[IndexSymbol, IndexSequence]:
+    """
+    This function takes the index sequence for a global read and removes all
+    workgroup and induction level indexing. This is necessary for writes to shared memory
+    that operate on promoted memory.
+    """
+    workgroup_ids = [WORKGROUP_0, WORKGROUP_1, WORKGROUP_2]
+    new_index = {key: index[key].subs({w: 0 for w in workgroup_ids}) for key in index}
+    for key in new_index:
+        for constraint in tilingConstraints:
+            new_index[key] = new_index[key].subs({constraint.induction_var: 0})
+    return new_index
+
+
+def materialize_shape(
+    constraint_tile_size: dict[IndexSymbol, int], symbolic_shape: list[IndexSymbol]
+) -> list[int]:
+    materialized_shape = []
+    idxc = IndexingContext.current()
+    for dim in symbolic_shape:
+        if dim in constraint_tile_size:
+            materialized_shape.append(constraint_tile_size[dim].subs(idxc.subs))
+        else:
+            materialized_shape.append(dim.subs(idxc.subs))
+    return materialized_shape
+
+
+def identify_optimizable_loads(
+    global_read_nodes: list[fx.Node],
+    constraint_tile_size: dict[IndexSymbol, int],
+    max_elements_per_load: int,
+) -> list[Read]:
+    """
+    Identify sub-optimal global loads that can be removed. A given memory has
+    sub-optimal global loads if
+        num_global_loads > (M * N) / (T * L)
+    where the memory has shape [M, N], there are T threads and each thread can load L elements.
+    """
+    optimizable_loads: dict[fx.Node, tuple[int, Read]] = {}
+    for read_node in global_read_nodes:
+        custom = get_custom(read_node)
+        if custom.memory in optimizable_loads:
+            continue
+        materialized_shape = materialize_shape(
+            constraint_tile_size, custom.type.symbolic_shape
+        )
+        # Ensure that the innermost dimension of the shape is a multiple of the elements being loaded.
+        if materialized_shape[-1] % max_elements_per_load == 0:
+            continue
+
+        total_number_of_elements = prod(materialized_shape)
+        expected_number_of_loads = total_number_of_elements // max_elements_per_load
+        actual_number_of_loads = len(
+            [x for x in global_read_nodes if get_custom(x).memory == custom.memory]
+        )
+        if expected_number_of_loads >= actual_number_of_loads:
+            continue
+        optimizable_loads[custom.memory] = (expected_number_of_loads, custom)
```

Comment:
Should we also check that it's only users are write to shared memory? This will be helpful in the case where we are doing computation without trips to shared memory.

 For example, if we had a simple elementwise add with suboptimal global reads/tile size, we might cause compilation failure or numerical issues.